{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GC[ATGC]GC\n",
      "{'AAT': 1, 'ATG': 1, 'ATC': 2, 'ACG': 1, 'TAC': 1, 'TGA': 2, 'TCG': 2, 'GAT': 2, 'GTA': 1, 'GCT': 1, 'CTG': 1, 'CGA': 1, 'CGT': 1, 'CGC': 1}\n",
      "count for TGA is 2\n",
      "count for AAA is 0\n",
      "count for GTA is 1\n",
      "count for TTT is 0\n",
      "ATC\n",
      "TGA\n",
      "TCG\n",
      "GAT\n",
      "dict_keys(['AAT', 'ATG', 'ATC', 'ACG', 'TAC', 'TGA', 'TCG', 'GAT', 'GTA', 'GCT', 'CTG', 'CGA', 'CGT', 'CGC'])\n",
      "ATC\n",
      "GAT\n",
      "TCG\n",
      "TGA\n",
      "ATC\n",
      "TGA\n",
      "TCG\n",
      "GAT\n"
     ]
    }
   ],
   "source": [
    "#In-chapter exercises for Chapter 8 - Dictionaries:\n",
    "\n",
    "#Storing paired data is an incredibly common problem that must be addressed in programming.\n",
    "#For example, we may want to store paired items like (1) codons and their associated amino \n",
    "#acid residues, (2) colleagues' names and their email addresses, (3) words and their \n",
    "#definitions, etc. In each of these examples, we have a key and a value. For example 1, our\n",
    "#key could be called \"codon\" and its value \"amino acid residue.\" In example 2, they key is\n",
    "#\"name\" and the value \"email address\"; in example 3, key = \"word\" and value = \"definition\".\n",
    "#For the last example, we have a tool for storing words and definitions in Python called a \n",
    "#dictionary.\n",
    "\n",
    "#Creating a dictionary involves curly brackets. Each pair of data, consisting of a key and\n",
    "#a value, is called an \"item\". Each item is separated with a comma; keys and values are \n",
    "#separated with colons. We can write a dictionary on a single line, like so:\n",
    "    #enzymes = {'EcoRI':r'GAATTC', 'AvaII':r'GG(A|T)CC', 'BisI':r'GC[ATGC]GC'}\n",
    "#But splitting a dictionary definition over several lines makes it easier to read:\n",
    "enzymes = {\n",
    "    'EcoRI' : r'GAATTC',\n",
    "    'AvaII' : r'GG(A|T)CC',\n",
    "    'BisI' : r'GC[ATGC]GC'\n",
    "}\n",
    "#This is how you get a bit of info from a dictionary. It's similar to getting an item from a\n",
    "#list, but you don't use indexes. You provide the key and the value:\n",
    "print(enzymes['BisI'])\n",
    "#Here's how you would create an empty dictionary and add elements to it:\n",
    "tRNA_codons = {}\n",
    "tRNA_codons['alanine'] = r'GC[ATGC]'\n",
    "tRNA_codons['tryptophan'] = r'TGG'\n",
    "tRNA_codons['phenylalanine'] = r'TT[TC]'\n",
    "#Here's how you delete a key from a dictionary:\n",
    "tRNA_codons.pop('alanine')\n",
    "#Let's make a dictionary of counts for a DNA sequence:\n",
    "dna = \"AATGATCGATCGTACGCTGA\"\n",
    "counts = {}\n",
    "for base1 in ['A', 'T', 'G', 'C']:\n",
    "    for base2 in ['A', 'T', 'G', 'C']:\n",
    "        for base3 in ['A', 'T', 'G', 'C']:\n",
    "            trinucleotide = base1 + base2 + base3\n",
    "            count = dna.count(trinucleotide)\n",
    "            if count > 0:\n",
    "                counts[trinucleotide] = count\n",
    "print(counts)\n",
    "#This is great, but now we can't look up counts for codons that are 0, because we only wanted\n",
    "#things in our dictionary with a count of 1 or greater. We will get an error if we try to print\n",
    "#any of those counts. We can test for whether something exists in a dictionary by using 'get'.\n",
    "#In the example below, 0 is the default value, which will be overridden if the item exists in\n",
    "#the counts dictionary:\n",
    "print(\"count for TGA is \" + str(counts.get('TGA', 0)))\n",
    "print(\"count for AAA is \" + str(counts.get('AAA', 0)))\n",
    "print(\"count for GTA is \" + str(counts.get('GTA', 0)))\n",
    "print(\"count for TTT is \" + str(counts.get('TTT', 0)))\n",
    "\n",
    "#We can iterate over a dictionary to do something to all of our items. This way of doing so,\n",
    "#in order to list all codons with a count of 2, is kind of inefficient:\n",
    "for base1 in ['A', 'T', 'G', 'C']:\n",
    "    for base2 in ['A', 'T', 'G', 'C']:\n",
    "        for base3 in ['A', 'T', 'G', 'C']:\n",
    "            trinucleotide = base1 + base2 + base3\n",
    "            if counts.get(trinucleotide, 0) == 2:\n",
    "                print(trinucleotide)\n",
    "#So let's use the 'keys' method instead. This will print a list of all the keys in our counts\n",
    "#dictionary:\n",
    "print(counts.keys())\n",
    "#Here's how we can list all keys with a count of 2 more concisely. The \"sorted\" text sorts\n",
    "#our dictionary so we get consistently ordered results. Otherwise, since dictionaries are\n",
    "#inherently unordered, we will get a different order in our output every time:\n",
    "for trinucleotide in sorted(counts.keys()):\n",
    "    if counts.get(trinucleotide) == 2:\n",
    "        print(trinucleotide)\n",
    "#We can also iterate over items in a dictionary:\n",
    "for trinucleotide, count in counts.items():\n",
    "    if count == 2:\n",
    "        print(trinucleotide)\n",
    "#This will return a list of pairs of values, something we haven't seen before."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#End-of-chapter exercises:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ASMFADASEQRNGRSHLALIL*AGQIDSGGNLR\n"
     ]
    }
   ],
   "source": [
    "#DNA Translation:\n",
    "#Write a program that will translate a DNA sequence into protein. Your program\n",
    "#should use the standard genetic code which can be found at the provided NCBI URL.\n",
    "\n",
    "#Generate a random DNA sequence:\n",
    "sequence = 'gctagtatgttcgcggacgcgtcagaacaaagaaatggccgctcgcacttggcgctcatactctaagctggtcagattgattcaggggggaatctacgaa'\n",
    "dna = sequence.upper()\n",
    "\n",
    "#Make a dictionary for codon translations:\n",
    "translations = {\n",
    "    'TTT':'F', 'TCT':'S', 'TAT':'Y', 'TGT':'C', 'TTC':'F', 'TCC':'S', 'TAC':'Y', 'TGC':'C',\n",
    "    'TTA':'L', 'TCA':'S', 'TAA':'*', 'TGA':'*', 'TTG':'L', 'TCG':'S', 'TAG':'*', 'TGG':'W',  \n",
    "    'CTT':'L', 'CCT':'P', 'CAT':'H', 'CGT':'R', 'CTC':'L', 'CCC':'P', 'CAC':'H', 'CGC':'R', \n",
    "    'CTA':'L', 'CCA':'P', 'CAA':'Q', 'CGA':'R', 'CTG':'L', 'CCG':'P', 'CAG':'Q', 'CGG':'R', \n",
    "    'ATT':'I', 'ACT':'T', 'AAT':'N', 'AGT':'S', 'ATC':'I', 'ACC':'T', 'AAC':'N', 'AGC':'S',\n",
    "    'ATA':'I', 'ACA':'T', 'AAA':'K', 'AGA':'R', 'ATG':'M', 'ACG':'T', 'AAG':'K', 'AGG':'R', \n",
    "    'GTT':'V', 'GCT':'A', 'GAT':'D', 'GGT':'G', 'GTC':'V', 'GCC':'A', 'GAC':'D', 'GGC':'G', \n",
    "    'GTA':'V', 'GCA':'A', 'GAA':'E', 'GGA':'G', 'GTG':'V', 'GCG':'A', 'GAG':'E', 'GGG':'G'}\n",
    "\n",
    "#Write a function to do translate DNA:\n",
    "def translate_dna(sequence):\n",
    "    sequence = sequence.upper()\n",
    "    last_codon_start = len(sequence) - 2\n",
    "    protein = \"\"\n",
    "    for start in range(0, last_codon_start, 3):\n",
    "        end = start + 3\n",
    "        codon = sequence[start:end]\n",
    "        amino_acid = translations.get(codon, 'X') #The 'X' is a default, which will be returned \n",
    "        #if there is an unknown 'N' base in our DNA sequence, which will create a key that \n",
    "        #doesn't exist in our dictionary. 'X' thus stands for an unknown amino acid residue.\n",
    "        protein = protein + amino_acid\n",
    "    return protein\n",
    "    \n",
    "#Translate DNA to test if the function works:\n",
    "print(translate_dna(dna))\n",
    "\n",
    "#Let's assert some things just to make sure:\n",
    "assert(translate_dna('ACTACTACTACT')) == 'TTTT'\n",
    "assert(translate_dna('actactactact')) == 'TTTT'\n",
    "assert(translate_dna('actactactactnnnnnanaaaaa')) == 'TTTTXXXK'"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
